### PR TITLE
docs: Add a provider framework guide.

### PR DESCRIPTION
**Pull request by paddycarver**
Document how to write a new provider for Terraform, or contribute new
pieces or functionality to an existing provider.

This doc should be largely feature complete. Things I would have liked
to included that didn't make it into this cut:

* How to make resources importable
* How to test resources, data sources, migrations, importing, etc.
* A better explanation of the configuration of a provider, separating
  out the Schema property and the ConfigureFunc more clearly.
* A better diagram on the lifecycle page.
* How to contribute documentation for your provider to the website.
* Useful practices for log messages.
* Notes on error wrapping
* Using timeouts
* Where, _exactly_, Exists fits into the lifecycle and when it should be used

While these things would be great to have, I'm calling this the MVP,
with an intent to add that information over time (or other people can PR
it in). For now, I think what we have is valuable enough to get it
published.

Because of the size of this PR, and the number of people involved in reviewing
it, it'd be great if reviews could be restricted to just the things that should
block a merge. Suggestions for enhancements or future directions are welcome,
but it'd be great to keep those as comments. Just keep in mind that changes
require people to re-review, that's all. :)

### CODE REVIEW

File: website/source/layouts/docs.erb [hunk 1]
```diff
@@ -552,6 +552,38 @@
           <li<%= sidebar_current("docs-internals-plugins") %>>
             <a href="/docs/internals/internal-plugins.html">Internal Plugins</a>
           </li>
+					<li<%= sidebar_current("docs-internals-provider-guide") %>>
+            <a href="/docs/internals/providers/index.html">Writing Your Own Provider</a>
+            <ul class="nav">
+              <li<%= sidebar_current("docs-internals-provider-guide-lifecycle") %>>
+                <a href="/docs/internals/providers/lifecycle.html">Understanding the Terraform Lifecycle</a>
```
Comment (sethvargo) [inline website/source/layouts/docs.erb]: This is really long for the sidebar. I'd say "Terraform Lifecycle"

File: website/source/layouts/docs.erb [hunk 2]
```diff
@@ -552,6 +552,38 @@
...
+              </li>
+              <li<%= sidebar_current("docs-internals-provider-guide-vs-resources") %>>
+                <a href="/docs/internals/providers/providers-v-resources.html">Providers vs. Resources & Data Sources</a>
```
Comment (sethvargo) [inline website/source/layouts/docs.erb]: Also, this is a longish title. Can we shorten?

File: website/source/layouts/docs.erb [hunk 3]
```diff
@@ -552,6 +552,38 @@
...
+              </li>
+              <li<%= sidebar_current("docs-internals-provider-guide-new-provider") %>>
+                <a href="/docs/internals/providers/new-provider.html">Creating a New Provider</a>
```
Comment (sethvargo) [inline website/source/layouts/docs.erb]: I think these can be "Creating Providers" and "Creating Resources" for brevity.

File: website/source/layouts/docs.erb [hunk 4]
```diff
@@ -552,6 +552,38 @@
           <li<%= sidebar_current("docs-internals-plugins") %>>
             <a href="/docs/internals/internal-plugins.html">Internal Plugins</a>
           </li>
+					<li<%= sidebar_current("docs-internals-provider-guide") %>>
```
Comment (paddycarver) [inline website/source/layouts/docs.erb]: Hm, weird, thought I caught all those. Thanks! I'll fix that.

File: website/source/layouts/docs.erb (hunk 2 above)
Comment (paddycarver) [inline website/source/layouts/docs.erb]: I couldn't find a way to shorten this, and looking at the section, it wasn't really... saying much. I've just deleted it and the link in the sidebar.

**Comment by paddycarver** [comment]
Thanks for the speedy review @sethvargo. I'm pretty sure I addressed your concerns, and tried to apply them consistently.

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+layout: "docs"
+page_title: "Creating Data Sources"
+sidebar_current: "docs-internals-provider-guide-new-data-source"
+description: |-
+  How to get started adding a data source to an existing provider.
+---
+
+# Creating Data Sources
+
+Once we have a Provider, we can add some Data Sources to that provider, which
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: There's a lot of "we" and "you" here. You are writing technical guides, so this is generally frowned upon.

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+Inside this file, you’re going to define a function that takes no arguments and
+returns a `*schema.Resource`, just like Resources.  While you can technically
+name your function anything, we recommend (especially if you plan on
+contributing your code back to the Terraform repo!) that you name it the
+camelCase version of the file name. For example, `data_source_aws_eip.go` would
+contain `dataSourceAwsEip`. We’ll be defining a _lot_ of functions in a
+Provider, and this kind of naming scheme really helps keep things organised and
+easy to find.
+
+## Registering the Data Source  With the Provider
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: There's an extra space after "Souce" and "With" should not be capitalized

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+we have a file naming guideline: we use
+`data_source_{api}_{data_source_name}.go` as the template to name our files,
+and tend to stick to  single Data Source per file. For example, AWS has
+`data_source_aws_acm_certificate.go`, because it reads certificates from AWS’
+ACM service. Some providers only have a single API, so we use the provider name
+in that case. For example, Docker has `data_source_docker_registry_image.go`.
+If you want to contribute your code back to the Terraform repo, it makes things
+easier if you follow this convention.
+
+Inside this file, you’re going to define a function that takes no arguments and
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: Do we have code samples for this? It's really challenging to follow along the walls of text without some copy-paste examples or sample code.

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+be the name of the Data Source used in state and configuration files, and (by
+convention) should match the `{data_source_name}` part used in the filename for
+the file that defines the Resource. Nothing will necessarily break if you don’t
+do this, but if you happen to accidentally have a collision with any other
+resource in any other provider, things will likely break. This convention helps
+keep things neat and orderly while avoiding conflicts.
+
+## Defining the Data Source Properties
+
+Each Data Source has a set of properties called the “schema” that is stored in
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: Fancy quotes. Fancy quotes. Fancy quotes. Please do a mass find-and-replace. These will make the builder sad and is a by-product of working in a Google doc. You can disable fancy quotes in your google settings.

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+contributing your code back to the Terraform repo!) that you name it the
+camelCase version of the file name. For example, `data_source_aws_eip.go` would
+contain `dataSourceAwsEip`. We’ll be defining a _lot_ of functions in a
+Provider, and this kind of naming scheme really helps keep things organised and
+easy to find.
+
+## Registering the Data Source  With the Provider
+
+Now that you’ve got your function, switch back to your Provider definition, and
+add the Data Source to the Provider’s `DataSourcesMap` property. The key will
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: I find it weird that we keep capitalizing "Data Sources" and "Resources"

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+and so on.
+
+These properties are defined in the `Schema` property of the Data Source
+definition. It takes a map with the keys being the property name and the values
+being `*Schema` structs. The `*Schema` structs define some type information
+(what kind of data to expect, etc.) along with some [advanced
+behaviour](/docs/internals/providers/schema.html) for resources that helps
+Terraform do the right thing without you needing to write a bunch of code.
+
+~> **Note:** “id” is a reserved property name. Don’t call your property “id”.
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: Another fancy quote, this time it's the single quote in "Don't".

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+ResourceData](/docs/internals/providers/resource-data.html). For Data Sources,
+the `*ResourceData` struct is used to pull the identifying information
+necessary to retrieve the information requested from the API, and as a place to
+store the information the API returns.
+
+The `interface{}` is the same `interface{}` returned by the Provider’s
+`ConfigureFunc`. Typically, this is where you’d put the configured API client,
+for example.
+
+The `Read` function should then use the passed `*ResourceData` to retrieve the
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: PV

File: website/source/docs/internals/providers/new-provider.html.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+layout: "docs"
+page_title: "Creating Providers"
+sidebar_current: "docs-internals-provider-guide-new-provider"
+description: |-
+  How to get started building a new provider.
+---
+
+# Creating Providers
+
+Say you’ve found a new API provider that Terraform doesn’t support, and you
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-provider.html.md]: This isn't really in line with the technical documentation we author.

File: website/source/docs/internals/providers/new-provider.html.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+methods, but the good news is it’s not your job to call or define _any_ of
+them. You can safely ignore them.
+
+## Creating the Package
+
+To define a new provider, you need to create a new package. If you are
+contributing this to the Terraform codebase, create a directory named after
+your plugin in the `builtin/providers` directory of Terraform. If you’re
+planning on maintaining it outside of Terraform as a standalone plugin, create
+a new repository for it-we recommend
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-provider.html.md]: You'll still want spaces around these dashes - the markdown parser will convert them

File: website/source/docs/internals/providers/new-provider.html.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+`interface{}` or an `error`. If you don’t return an error, the interface you
+return will be stored and passed to all your resources as a meta parameter. If
+you do return an error, that will bubble up to the Terraform CLI and abort
+whatever command is being run. **Note:** please return errors instead of using
+`panic` or `os.Exit`.
+
+## Registering Your Provider
+
+Because Terraform uses a plugin architecture, we need to register our new
+plugin with Terraform before Terraform knows how to use it. (Fun fact: even the
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-provider.html.md]: Remove the fun fact

File: website/source/docs/internals/providers/new-resource.html.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+anything, we recommend (especially if you plan on contributing your code back
+to the Terraform repo!) that you name it the camelCase version of the file
+name. For example, `resource_compute_instance.go` would contain
+`resourceComputeInstance`. We’ll be defining a _lot_ of functions in a
+Provider, and this kind of naming scheme really helps keep things organised and
+easy to find.
+
+## Registering the Resource With the Provider
+
+Now that you’ve got your function, switch back to your Provider definition, and
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: As I said above, I don't think type of language is appropriate for technical documentation or guides.

Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: Provider probably shouldn't be capitalized either.

File: website/source/docs/internals/providers/new-data-source.html.md (hunk 7 above)
Comment (paddycarver) [inline website/source/docs/internals/providers/new-data-source.html.md]: I don't have code samples at the moment, and do not believe I'd be able to come up with samples of the appropriate quality to meet the business timeline for publishing this. (~sometime next week, earlier the better).

**Comment by paddycarver** [comment]
Thanks for the in-depth feedback, @sethvargo, really appreciate it. I've addressed most your points, I believe--hyphens should now have spaces, magic quotes have been eliminated, paragraphs have become lists, images have been added.

There are a lot of voice concerns here and they're all valid and I want to address them all. However, because this is a writing voice I don't have much practice in, I've updated the following files to try and make the changes you've requested. Could you just take a look at those (ignoring the voice in the others for the moment) and let me know if you're happy with those changes? If so, I'll update the rest. I just want to be sure I've got this right before I spend more time applying it.

The files with updated voice:

* async.html
* index.html
* lifecycle.html
* new-data-source.html

File: website/source/docs/internals/providers/index.html.md [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+   resources.
+
+By dividing things like this, it is much easier to contribute new functionality
+or new APIs to Terraform: contributtors can essentially just trust the core
+section to do its job, and only worry about their resources.
+
+One way to think of the separation is that the core tells the providers what to
+do, and the providers make that change against the APIs.
+
+To hook into this behaviour, Terraform has a provider framework. This guide
```
Comment (sethvargo) [inline website/source/docs/internals/providers/index.html.md]: This is the "british" spelling of "behavior". I think we should be consistent and use the "american" one instead.

File: website/source/docs/internals/providers/lifecycle.html.md [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+layout: "docs"
+page_title: "Terraform Lifecycle"
+sidebar_current: "docs-internals-provider-guide-lifecycle"
+description: |-
+  Information about the lifecycle your provider fits into.
+---
+
+# Terraform Lifecycle
+
+Terraform runs follow a predictable lifecycle:
```
Comment (sethvargo) [inline website/source/docs/internals/providers/lifecycle.html.md]: What is a run?

(I know, but it'd be helpful to provide a bit more context here, since we usually say "plan" or "apply" exlicitly)

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+layout: "docs"
+page_title: "Creating Data Sources"
+sidebar_current: "docs-internals-provider-guide-new-data-source"
+description: |-
+  How to get started adding a data source to an existing provider.
+---
+
+# Creating Data Sources
+
+Data sources are components of providers that pull information from outside of
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: This isn't necessarily true. Data sources pull data from external sources, but those externals sources could be managed by Terraform too. This makes it sound mutually exclusive.

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+Terraform's managed resources, to make it available in configuration.  They
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: There is a double space after period here.

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+should exist in the same package as the provider.
+
+The built-in plugins for Terraform that are shipped as part of the codebase
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: This reads choppy. I would remove "that are shipped as part of the codebase" or move it to its own sentence.

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+have a file naming guideline: they use
+`data_source_{api}_{data_source_name}.go` as the template to name files, and
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: We use the convention `<THING>` in these cases. So `data_source_<API>_<DATA_SOURCE_NAME>` instead

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+tend to stick to  single data source per file. For example, AWS has
+`data_source_aws_acm_certificate.go`, because it reads certificates from AWS'
+ACM service. Some providers only have a single API, so they use the provider
+name in that case. For example, Docker has
+`data_source_docker_registry_image.go`. Following these conventions makes it
+easier to contribute code back to the Terraform codebase.
+
+Just like resources, data sources are functions that take no arguments and
+return a `*schema.Resource`. While functions can technically be named anything,
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: Remove "technically"

File: website/source/docs/internals/providers/new-data-source.html.md [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+ResourceData](/docs/internals/providers/resource-data.html). Data sources use
+the `*schema.ResourceData` struct to pull the identifying information necessary
+to retrieve the information requested from the API, and as a place to store the
+information the API returns.
+
+The `interface{}` is the same `interface{}` returned by the
+`*schema.Provider`'s `ConfigureFunc`. It generally contains a configured API
+client or similar form of access for the provider.
+
+The `Read` function should use the passed `*ResourceData` to retrieve the ID
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-data-source.html.md]: should use -> uses

File: website/source/docs/internals/providers/new-provider.html.md [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+layout: "docs"
+page_title: "Creating Providers"
+sidebar_current: "docs-internals-provider-guide-new-provider"
+description: |-
+  How to get started building a new provider.
+---
+
+# Creating Providers
+
+Say you've found a new API provider that Terraform doesn't support, and you
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-provider.html.md]: There's still a lot of first person in this entire page.

File: website/source/docs/internals/providers/new-resource.html.md [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+---
+
+# Creating Resources
+
+Now that you've got a Provider, it's time to add some Resources to that
+Provider to make it actually _do_ something. To do this, you need to write some
+code in the package your Provider is defined in.
+
+In the built-in plugins for Terraform that are shipped as part of the codebase,
+we have a file naming guideline: we use `resource_{resource_name}.go` as the
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: `<THING>`

File: website/source/docs/internals/providers/new-resource.html.md [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+anything, we recommend (especially if you plan on contributing your code back
+to the Terraform repo!) that you name it the camelCase version of the file
+name. For example, `resource_compute_instance.go` would contain
+`resourceComputeInstance`. We'll be defining a _lot_ of functions in a
+Provider, and this kind of naming scheme really helps keep things organised and
+easy to find.
+
+## Registering the Resource With the Provider
+
+Now that you've got your function, switch back to your Provider definition, and
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: "You've got your"

File: website/source/docs/internals/providers/new-resource.html.md [hunk 27]
```diff
@@ -0,0 +1,127 @@
...
+add the Resource to the Provider's `ResourcesMap` property. The key will be the
+name of the Resource used in state and configuration files, and (by convention)
+should match the `{resource_name}` part used in the filename for the file that
+defines the Resource. Nothing will necessarily break if you don't do this, but
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: "Nothing will... if you don't" - double negative and challenging to follow

File: website/source/docs/internals/providers/new-resource.html.md [hunk 28]
```diff
@@ -0,0 +1,127 @@
...
+if you happen to accidentally have a collision with any other resource in any
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: happen to accidentally -> accidentally

Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: have a collision -> collide

File: website/source/docs/internals/providers/new-resource.html.md [hunk 29]
```diff
@@ -0,0 +1,127 @@
...
+being `*Schema` structs. The `*Schema` structs define some type information
+(what kind of data to expect, etc.) along with some [advanced
+behaviour](/docs/internals/providers/schema.html) for resources that helps
+Terraform do the right thing without you needing to write a bunch of code.
+
+~> **Note:** "id" is a reserved property name. Don't call your property "id".
+
+## Calling the API
+
+Now that we have a Resource fully defined, it's time to make it do something.
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: Now that the resource is fully defined, it's time to add functionality to manage upstream resources.

File: website/source/docs/internals/providers/new-resource.html.md [hunk 30]
```diff
@@ -0,0 +1,127 @@
...
+
+The `*ResourceData` struct contains the state of the Resource as it should be.
+It's an amalgamation of several different sources of data, which are explained
+further in [Using ResourceData](/docs/internals/providers/resource-data.html).
+
+The `interface{}` is the same `interface{}` returned by the Provider's
+`ConfigureFunc`. Typically, this is where you'd put the configured API client,
+for example.
+
+Of all these functions, only the `Update` and `Exists` functions are optional.
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: There are definitely other functions on the helper/schema - `Importer`, `Timeouts`, `MigrateState`

File: website/source/docs/internals/providers/new-resource.html.md [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+If the `Update` function isn't set, the Resource will be treated as something
+that cannot be updated. If the `Exists` function isn't set, Terraform will not
+check if the Resource already/still exists.
+
+The `Create` function should use the passed `*ResourceData` to make an API call
```
Comment (sethvargo) [inline website/source/docs/internals/providers/new-resource.html.md]: Reading this (as someone who has written a lot of Terraform), I'm still not really sure what I do with this information. I think adding code examples and method signatures here is very important.

File: website/source/docs/internals/providers/resource-data.html.md [hunk 32]
```diff
@@ -0,0 +1,180 @@
...
+
+When working with the `Create`, `Read`, `Update`, `Destroy`, and `Exists`
+methods on your Provider, it's almost impossible to not run into
+`ResourceData`. It's a type that is used all over, so it's important to
+understand it. But `ResourceData` can also be tricky, because it's not
+abstracting a single logical concept in Terraform.
+
+## How to Conceptualise `ResourceData`
+
+The most useful way to think of `ResourceData` is not as a placeholder for any
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: I read this sentence 3 times and I'm still not sure what is being said. Could you rephrase or break into separate sentences please?

File: website/source/docs/internals/providers/resource-data.html.md [hunk 33]
```diff
@@ -0,0 +1,180 @@
...
+one concept in Terraform—it's not just information about the config or the
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: Space around `-`

File: website/source/docs/internals/providers/resource-data.html.md [hunk 34]
```diff
@@ -0,0 +1,180 @@
...
+
+To get practical, `ResourceData` abstracts four separate concepts into a single
+source of truth:
+
+* The current state
+* The config file
+* The diff being applied
+* Any calls to `ResourceData.Set`
+
+Those are expressed in the order of their priority. That is, the config
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: Remove `that is`

File: website/source/docs/internals/providers/resource-data.html.md [hunk 35]
```diff
@@ -0,0 +1,180 @@
...
+overrides any values set in the state, and every time you call `Set`, it
+overrides everything else.
+
+`ResourceData` often gets confused for either the state or the config, but it's
+important to realise that it's a step removed from these concepts, and is used
+more as an understanding of what the infrastructure _should_ be.
+
+## Retrieving Properties
+
+To retrieve properties from `ResourceData`, use the `Get` or `GetOk` methods.
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: Again, I see lots of text, but what is `GetOk`? How do I use it? Adding a 2-3 line code sample here would do wonders for the usefulness of this guide.

File: website/source/docs/internals/providers/resource-data.html.md [hunk 36]
```diff
@@ -0,0 +1,180 @@
...
+`Get` takes a property name or address (e.g., `myprop.0.value`) and returns it
+as an `interface{}`. It's important to note that the framework will always
+guarantee consistency about the underlying type of that `interface{}`. If the
+key doesn't exist in your schema, `Get` returns `nil`. If the key exists in
+your schema, but not in the config, `Get` returns the empty value for that
+type.
+
+`ResourceData` also has a `GetOk` method that functions identically to its
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: if it functioned identically, it would return the same thing. It functions _simarily_

File: website/source/docs/internals/providers/resource-data.html.md [hunk 37]
```diff
@@ -0,0 +1,180 @@
...
+`Get` method, but with an extra return parameter. The new return parameter
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: s/new/extra

File: website/source/docs/internals/providers/resource-data.html.md [hunk 38]
```diff
@@ -0,0 +1,180 @@
...
+
+When there's a difference between what our infrastructure is and what we want
+it to be, we want to be able to see what changed and what it should be. To aid
+in this, `ResourceData` provides a `HasChange` method and a `GetChange` method.
+Each takes a property key, just like `Get`. `HasChange` returns `true` if
+there's a change to that property; the change could be from drift (someone or
+something modifying the infrastructure outside of Terraform) or from config
+changes. The `GetChange` method returns two values; a basic understanding is
+that the first value is what the property _used to be_, and the second value is
+what the property _should be changed to_.  A slightly more nuanced
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: There are two spaces after this period

File: website/source/docs/internals/providers/resource-data.html.md [hunk 39]
```diff
@@ -0,0 +1,180 @@
...
+so it can be removed.
+
+## Setting State
+
+At the end of your `Create`, `Get`, `Update`, and `Delete` functions, whatever
+is in your `ResourceData` object is set as state. To manipulate this, the `Set`
+method is provided. It takes the key of a property, just like `Get`, as an
+argument, and the value to set it to.
+
+It's important that all your properties get set using the `Set` method, as
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: "get set using the `Set` method". Please try to rephrase this using words that aren't also keywords in this context.

File: website/source/docs/internals/providers/resource-data.html.md [hunk 40]
```diff
@@ -0,0 +1,180 @@
...
+infrastructure get persisted using the `Set` method.
+
+When using the `Set` method, you shouldn't dereference pointer values, as doing
+so incorrectly (for example, when they're set to a nil value) can crash
+Terraform. The `Set` method is capable of safely derefencing on its own, so the
+safest course of action is to just pass it the pointer.
+
+## Partial Updates
+
+Some resources and data sources cannot be managed using only a single API call;
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: This isn't a semicolon kind of thing. These are two separate thoughts.

File: website/source/docs/internals/providers/resource-data.html.md [hunk 41]
```diff
@@ -0,0 +1,180 @@
...
+for example, a resource may need its permissions set or a compute instance may
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: its' permissions (permission belongs to the resource)

File: website/source/docs/internals/providers/resource-data.html.md [hunk 42]
```diff
@@ -0,0 +1,180 @@
...
+need a disk attached to it, even though the instance and disk are modeled as a
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: drop "to it"

File: website/source/docs/internals/providers/resource-data.html.md [hunk 43]
```diff
@@ -0,0 +1,180 @@
...
+Usually, when a resource is created or updated, Terraform
+[builds](#what-resourcedata-abstracts) the `*ResourceData` from the current
+state, the config file, and the diff being applied. It then calls your `Create`
+or `Update` function (as appropriate) and any calls to the `*ResourceData`'s
+`Set` method get layered on top of the `*ResourceData`. After _successful_
+completion of the function, the `*ResourceData` gets stored as the new state.
+Should the function return an error, the `*ResourceData` gets discarded and is
+_not_ set as the state.
+
+The `Partial` method accepts a boolean designating whether partial state mode
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: Again, I really do not think this guide is useful without code samples. You're basically describing the method signature in a wall of text that could easily be replaced by a code sample of 5-6 lines with a comment.

File: website/source/docs/internals/providers/resource-data.html.md [hunk 44]
```diff
@@ -0,0 +1,180 @@
...
+
+Terraform special-cases the `id` property on resources. Do not use `id` as a
+key for a property. Terraform uses that internally, and trying to access an
+`id` property with the `Get` method will always return a zero value.
+
+To retrieve the ID, use the `GetId` method on `ResourceData`. To set it, use
+the `SetId` method, passing the string to use as an ID. Only strings may be
+used as IDs.
+
+You should always set the ID as soon as you possibly can. Even with
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resource-data.html.md]: "as soon as you possibly can" -> as soon as possible

File: website/source/docs/internals/providers/resources-vs-data-sources.html.md [hunk 45]
```diff
@@ -0,0 +1,21 @@
...
+layout: "docs"
+page_title: "Resources vs. Data Sources"
+sidebar_current: "docs-internals-provider-guide-resources-v-data-sources"
+description: |-
+  Understanding whether to use a resource or a data source.
+---
+
+# Resources vs Data Sources
+
+Resources and Data Sources are incredibly similar, and for the most part, can
```
Comment (sethvargo) [inline website/source/docs/internals/providers/resources-vs-data-sources.html.md]: s/incredibly/very

File: website/source/docs/internals/providers/schema.html.md [hunk 46]
```diff
@@ -0,0 +1,322 @@
...
+information you'd expect: the type of information (a string, an integer, a
+list, etc.) that it expects, whether that information is optional or computed,
+etc.; but it also has a bunch of special features that can be defined for each
+property.
+
+## Working With Computed
+
+There are some fields on an Resource or Data Source that can't be influenced by
+the user at all, which the API provider is the source of truth for. Things like
+the timestamp a resource was created, or the ID of a resource. For these
```
Comment (sethvargo) [inline website/source/docs/internals/providers/schema.html.md]: This isn't a sentence

File: website/source/docs/internals/providers/new-provider.html.md (hunk 24 above)
Comment (paddycarver) [inline website/source/docs/internals/providers/new-provider.html.md]: Yeah, I mentioned I had only updated the first few pages. Sorry for the confusion, will try to be clearer next time. Was trying to save us both some time and spot-check for the voice you wanted before I rewrote a 7000 word document (or made you re-edit a 7000 word document).

File: website/source/docs/internals/providers/new-resource.html.md (hunk 30 above)
Comment (paddycarver) [inline website/source/docs/internals/providers/new-resource.html.md]: I'm not sure what this is suggesting. `Timeout` and `MigrateState` do not call the API, so I'm not sure they would belong under a section titled "Calling the API". `Importer` _could_ be considered part of this, but would likely make more sense in its own section, given its lifecycle is a bit... different. `MigrateState` is talked about in the `schema.html` page, and `Importer` and `Timeouts` were both punted on for this revision.

**Comment by paddycarver** [comment]
Closing in favour if #14250.
